client/src/main/java/io/pravega/client/auth/CredentialsHelper.java

```diff
@@ -77,10 +82,24 @@ private static Credentials extractCredentialsFromEnv() {
     }
 
     private static Credentials credentialFromMap(Map<String, String> retVal) {
+
+        String expectedMethod = retVal.get(AUTH_METHOD);
+
+        // Load the class dynamically if the user wants it to.
+        if (retVal.containsKey(AUTH_METHOD_LOAD_DYNAMIC) && Boolean.parseBoolean(retVal.get(AUTH_METHOD_LOAD_DYNAMIC))) {
+            ServiceLoader<Credentials> loader = ServiceLoader.load(Credentials.class);
+            for (Credentials creds : loader) {
+                if (creds.getAuthenticationType().equals(expectedMethod)) {
+                    return creds;
```

Comment:
Just a question: so in the case of a dynamically loaded `Credentials` implementation, that implementation doesn't have access to `retVal`.  So it's responsible for how it wants to acquire some parameters.  Probably through env and properties also.

That's probably ok/good.  This way the implementation isn't bound to any particular way of naming parameters.  

For example, say I have a FooCredentials class.  I set `pravega.client.auth.method` to `foo`, and `pravega.client.auth.loadDynamic` to true.     It loads FooCredentials.  Now FooCredentials itself may want to know some stuff like, where to find an IDP server (url), and username password that relate to that.

My understanding is it's up to `FooCredentials` how to acquire those (look for some FOO_USERNAME env or foo_username property or some file somewhere).   Is that correct?

Reply:
Yes. That understanding is correct. It is up to the individual implementation to get the details the way it wants.